src/web_peer_connection.cpp

```diff
@@ -649,14 +658,39 @@ void web_peer_connection::handle_redirect(int const bytes_left)
 		if (web->have_files.get_bit(file_index) == false)
 		{
 			web->have_files.set_bit(file_index);
+			torrent_info const& info = t->torrent_file();
+
+			// let's try to predict file location names.
+			// we should try to do it cause original request will be dropped
+			// and further requests will use exclusive range pick logic
+			// which exclude edge pieces file locations.
+
+			std::string original_path = escape_file_path(info.orig_files(), file_index);
+			std::size_t redirect_path_idx = redirect_path.rfind(original_path);
+			if (redirect_path_idx != std::string::npos) // we can predict new location
+			{
+				std::string redirect_path_prefix = redirect_path.substr(0, redirect_path_idx);
+				file_storage const& fs = info.files();
+				for (file_index_t fi(0); fi < fs.end_file(); ++fi)
+				{
+					if (info.orig_files().pad_file_at(fi))
+						continue;
+					if (web->redirects.find(fi) != web->redirects.end())
+						continue;
+					web->have_files.set_bit(fi);
```

Comment:
this is optimisitic assumption that all the rest of files will be at these new predicted locations (it's done only once if redirects[file_index] is not filled). if assumption is incorrect then have_file.clear_bit occured for this web_seed.
anyway it's the current only way to download unaligned files from plain redirect (new location for all files are equal by file_path string endings). without this prediction download simply stuck cause web_peer_connection::on_connect cut requests for edge pieces.